internal/wrapper/wrap_handler.go

```diff
@@ -156,3 +165,29 @@ func unmarshalEventForHandler(ev json.RawMessage, handler interface{}) (reflect.
 	}
 	return newMessage, err
 }
+
+func submitEnhancedMetrics(metricName string, ctx context.Context) {
```

Comment:
It might be cleaner if `submitEnhancedMetrics` and `getEnhancedMetrics` were methods on the MetricsListener. You could call them automatically from within HandlerStarted and HandlerFinished.

The only complication is that we'd have to change the signature of HandlerFinished so that it can check the value of `err` to see if it needs to write an errors metric. I don't think that would change very much, though.

Reply:
I like that idea!